triggers/daml/ClearingTrigger.daml

```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
                                 ((nextDeposit :: used), remaining)
   where
     nextAmount  = (.asset.quantity) . snd $ nextDeposit
+
+
+-- |Calculates the FairValue for all Derivatives known to Exchange
+handleCalculationRequests : Party -> ContractPair Clearing.FairValueCalculationRequest -> TriggerA ClearingState ()
+handleCalculationRequests party (requestCid, rq@Clearing.FairValueCalculationRequest{..}) = do
+  debug "handling calculation requests..."
+  allListings <- query @Listing.T
+  let optListingIdSet = Set.fromList <$> optListingIds
+  let listings = case optListingIdSet of
+        (Some ids) -> filterField (.listingId) (`Set.member` ids) allListings
+        None       -> allListings
+
+  forA_ listings (handleFairValueCalculation party currency upTo)
+  emitExerciseCmd requestCid Clearing.FairValueCalculationRequest_Ack
+  return ()
+
+-- |Calculates a FairValue for a single derivative.
+-- Order of precedence:
+--    1. The price of the most recent trade before the specified 'upTo'
```

Comment:
This doesn't sound right. Shouldn't you be taking price _after_ the specified date?

Reply:
`upTo` was how the requirement was given, the idea being "I would like the latest fair value, ignoring anything after 5pm". It might make sense to also have an optional `after : Time` to allow for a range.

Reply:
So what if
- the trade was at 9 am this morning? Will that be ignored because it's after 5 pm yesterday?
- the last trade was three days ago? (This happens in real life).

> upTo was how the requirement was given,

I'd like to understand if this was an internal requirement / Exberry / a client. Maybe we can chat off-line.

Reply:
This was an Eric requirement, but we can chat about it. For the first one, yes, if you specified 5pm yesterday then the trade was at 9am this morning will be ignored by design, since you're interested in doing calculations based on what happened yesterday. For the second, also yes based on this particular algorithm (which is overly simple and can be changed) since that is the most recent traded value.